File: Data Structures and Problem Solving Using Cpp 2nd/Josephus.cpp
#include "BinarySearchTree.h"


#ifdef USE_DOT_H
    #include <iostream.h>
#else
    #include <iostream>
    using namespace std;
#endif
    
#ifndef SAFE_STL
    #include <list>
#else
    #include "list.h"
    #include "StartConv.h"
#endif


// Return the winner in the Josephus problem.
// STL list implementation.
int jos1( int people, int passes )
{
    list<int> theList;
    list<int>::iterator itr;
    list<int>::iterator next;
    int i;

      // Construct the list.
    for( i = 1; i <= people; i++ )
        theList.push_back( i );

      // Play the game.
    for( itr = theList.begin( ); people-- != 1; itr = next )
    {
        for( i = 0; i < passes; i++ )
        {
            ++itr;                       // Advance
            if( itr == theList.end( ) )  // If past last player
                itr = theList.begin( );  // then go to first
        }

        next = itr;       // Maintain next node, for
        ++next;           // player who is after removed player 

        theList.erase( itr );            // Remove player

        if( next == theList.end( ) )     // Set next
            next = theList.begin( );
    }

    return *itr;     // Return player's number
}


// Recursively construct a perfectly balanced binary search
// tree by repeated insertions in O( N log N ) time.
void buildTree( BinarySearchTreeWithRank<int> & t, int low, int high )
{
    int center = ( low + high ) / 2;

    if( low <= high )
    {
        t.insert( center );
        buildTree( t, low, center - 1 );
        buildTree( t, center + 1, high );
    }
}

// Return the winner in the Josephus problem.
// Search tree implementation.
int jos2( int people, int passes )
{
    BinarySearchTreeWithRank<int> t;

    buildTree( t, 1, people );

    int rank = 1;
    while( people > 1 )
    {
        if( ( rank = ( rank + passes ) % people ) == 0 )
            rank = people;

        t.remove( t.findKth( rank ).get( ) );
        people--;
    }

    return t.findKth( 1 ).get( );
}

// Quickie main to test
int main( int argc, char **argv )
{
    if( argc == 3 )
    {
        cout << jos1( atoi( argv[ 1 ] ), atoi( argv[ 2 ] ) ) << endl;
        cout << jos2( atoi( argv[ 1 ] ), atoi( argv[ 2 ] ) ) << endl;
    }
    else
        cerr << "Usage: Jos People Passes" << endl;

    return 0;
}


#ifdef SAFE_STL
    #include "EndConv.h"
#endif

File: Data Structures and Problem Solving Using Java 4th/Josephus.java
import weiss.util.Iterator;
import weiss.util.Collection;
import weiss.util.LinkedList;
import weiss.util.ArrayList;

import weiss.nonstandard.BinarySearchTreeWithRank;

public final class Josephus
{
    /**
     * Return the winner in the Josephus problem.
     * Linked list implementation.
     * (Can replace with ArrayList or TreeSet).
     */
    public static int jos1( int people, int passes )
    {
        Collection<Integer> theList = new LinkedList<Integer>( );

            // Construct the list
        for( int i = 1; i <= people; i++ )
            theList.add( i );

            // Play the game;
        Iterator<Integer> itr = theList.iterator( );
        while( people-- != 1 )
        {
            for( int i = 0; i <= passes; i++ )
            {
                if( !itr.hasNext( ) )
                    itr = theList.iterator( );
         
                itr.next( );
            }               
            itr.remove( );    
        }
        
        itr = theList.iterator( );

        return itr.next( );
    }

    /**
     * Recursively construct a perfectly balanced BinarySearchTreeWithRank
     * by repeated insertions in O( N log N ) time.
     * t should be empty on the initial call.
     */
    public static void buildTree( BinarySearchTreeWithRank<Integer> t,
                                  int low, int high )
    {
        int center = ( low + high ) / 2;

        if( low <= high )
        {
            t.insert( center );

            buildTree( t, low, center - 1 );
            buildTree( t, center + 1, high );
        }
    }

    /**
     * Return the winner in the Josephus problem.
     * Search Tree implementation.
     */
    public static int jos2( int people, int passes )
    {
        BinarySearchTreeWithRank<Integer> t = new BinarySearchTreeWithRank<Integer>( );

        buildTree( t, 1, people );

        int rank = 1;
        while( people > 1 )
        {
            rank = ( rank + passes ) % people;
            if( rank == 0 )
                rank = people;

            t.remove( t.findKth( rank ) );
            people--;
        }

        return t.findKth( 1 );
    }


    // Quickie main to test

    public static void main( String [ ] args )
    {                
        try
        {
            if( args.length == 2 )
            {
                int arg1 = Integer.parseInt( args[ 0 ] );
                int arg2 = Integer.parseInt( args[ 1 ] );
                System.out.println( "JOS1: " + jos1( arg1, arg2 ) );
                System.out.println( "JOS2: " + jos2( arg1, arg2 ) );
            }
            else
                System.err.println( "Usage: Josephus People Passes" );
        }
        catch( NumberFormatException e )
          { System.err.println( "Usage: Josephus People Passes" ); }
    }
}